src/xeus_calc_interpreter.cpp

```diff
@@ -0,0 +1,178 @@
+        while (std::getline(input, token, ' '))
+        {
+            if (token.empty())
+            {
+                continue;
+            }
+
+            char c = token[0];
+            size_t idx = ops.find(c);
+
+            // check for operator
+            if (idx != std::string::npos)
+            {
+                while (!s.empty())
+                {
+                    int prec2 = s.top() / 2;
+                    int prec1 = idx / 2;
+                    if (prec2 > prec1 || (prec2 == prec1 && c != '^'))
+                    {
+                        ss << ops[s.top()] << ' ';
+                        s.pop();
+                    }
+                    else break;
+                }
+                s.push(idx);
+            }
+            else if (c == '(')
+            {
+                s.push(-2); // -2 stands for '('
+            }
+            else if (c == ')')
+            {
+                // until '(' on stack, pop operators.
+                while (s.top() != -2)
+                {
+                    ss << ops[s.top()] << ' ';
+                    s.pop();
+                }
+                s.pop();
+            }
+            else
+            {
+                ss << token << ' ';
+            }
+        }
+
+        while (!s.empty())
+        {
+            ss << ops[s.top()] << ' ';
+            s.pop();
+        }
+        publish_stream("stdout", "RPN = " + ss.str());
+        return ss.str();
+    }
+
+    double interpreter::compute_rpn(const std::string &expr)
+    {
+        publish_stream("stdout", "\nInput\tOperation\tStack after\n");
+        std::istringstream iss(expr);
+        std::vector<double> stack;
+        std::string token;
+        while (iss >> token)
+        {
+            publish_stream("stdout", token + "\t");
+            double tokenNum;
+            if (std::istringstream(token) >> tokenNum)
+            {
+                publish_stream("stdout","Push\t\t");
+                stack.push_back(tokenNum);
+            }
+            else
+            {
+                publish_stream("stdout", "Operate\t\t");
+                double secondOperand = stack.back();
+                stack.pop_back();
+                double firstOperand = stack.back();
+                stack.pop_back();
+                if (token == "*")
+                    stack.push_back(firstOperand * secondOperand);
+                else if (token == "/")
+                    stack.push_back(firstOperand / secondOperand);
+                else if (token == "-")
+                    stack.push_back(firstOperand - secondOperand);
+                else if (token == "+")
+                    stack.push_back(firstOperand + secondOperand);
+                else if (token == "^")
+                    stack.push_back(std::pow(firstOperand, secondOperand));
+                else
+                { //just in case
+                    publish_stream("stderr", "Error");
+                    std::exit(1);
```

Comment:
Don't exit, publish an error message on IOPub and return. Or you can throw an exception that has to be catched in `execute_request_impl`